design-system/packages/editor/src/EditorToolbar.tsx

```diff
@@ -0,0 +1,387 @@
+
+  return (
+    <PressResponder {...itemProps}>
+      <ToggleButton prominence="low" {...props} />
+    </PressResponder>
+  );
+}
+
+export function EditorToolbarSeparator() {
+  return <Divider orientation="vertical" flexShrink={0} />;
+}
+
+// =============================================================================
+// Utils
+// =============================================================================
+
+function filterDOMPropsWithLabelWarning<P extends AriaLabelingProps>(props: P) {
+  let { 'aria-labelledby': ariaLabelledby, 'aria-label': ariaLabel } = props;
+
+  if (!ariaLabelledby && !ariaLabel) {
+    console.warn(
+      'You must specify an aria-label or aria-labelledby attribute for accessibility.'
+    );
+  }
+
+  return filterDOMProps(props, { labellable: true });
+}
+
+function useToolbarItem<P extends { isDisabled?: boolean }>(props: P) {
+  let { isDisabled } = props;
+  let { state } = useToolbarContext();
+  let { lastFocusedId, setLastFocusedId } = state;
+  let id = useId();
+  let tabIndex = lastFocusedId === id || lastFocusedId == null ? 0 : -1;
+
+  // clear the last focused ID when the item is unmounted or becomes disabled,
+  // which will reset the tabIndex for each item to 0 avoiding a situation where
+  // the user cannot tab to any items
+  useEffect(() => {
+    let reset = (lastId: Key | null) => (lastId === id ? null : lastId);
+    if (isDisabled) {
+      setLastFocusedId(reset);
+    }
+    return () => {
+      setLastFocusedId(reset);
+    };
+  }, [id, isDisabled, setLastFocusedId]);
+
+  return {
+    itemProps: {
+      tabIndex,
+      id,
+      onFocus: () => {
+        setLastFocusedId(id);
+      },
+    },
+  };
+}
+
+function useToolbar(props: EditorToolbarProps, ref: RefObject<HTMLElement>) {
+  let [lastFocusedId, setLastFocusedId] = useState<Key | null>(null);
+  let { direction } = useLocale();
+  let focusManager = createFocusManager(ref, { wrap: true });
+  let isRtl = direction === 'rtl';
+
+  let onKeyDown = (e: KeyboardEvent) => {
+    if (!e.currentTarget.contains(e.target as HTMLElement)) {
+      return;
+    }
+
+    switch (e.key) {
+      case 'Home':
+        e.preventDefault();
+        e.stopPropagation();
+        focusManager.focusFirst();
+        break;
+      case 'End':
+        e.preventDefault();
+        e.stopPropagation();
+        focusManager.focusLast();
+        break;
+      case 'ArrowRight':
+      case 'ArrowDown':
+        e.preventDefault();
+        e.stopPropagation();
+        if (e.key === 'ArrowRight' && isRtl) {
+          focusManager.focusPrevious();
+        } else {
+          focusManager.focusNext();
+        }
+        break;
+      case 'ArrowLeft':
+      case 'ArrowUp':
+        e.preventDefault();
+        e.stopPropagation();
+        if (e.key === 'ArrowLeft' && isRtl) {
+          focusManager.focusNext({ wrap: true });
+        } else {
+          focusManager.focusPrevious({ wrap: true });
+        }
+        break;
```

Comment:
just a random idea: what if we supported the same keys as word navigation (option+left/right arrows on mac and i think ctrl+shift+left/right arrows on windows) to go between groups

Reply:
yeah, that could be nice. i'm not sure how to implement w/o skipping items that are outside of a group though 🤔

Reply:
i'd probably think of it as items outside of a group are essentially their own group? (just like words can be single characters)

Reply:
yeah, agreed. more of a technical concern; just need to think about how to filter the nodes to get the desired result